llama.cpp

```diff
@@ -2179,6 +2182,74 @@ static void llm_load_vocab(
     GGUF_GET_KEY(ctx, vocab.special_unk_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_UNK_ID));
     GGUF_GET_KEY(ctx, vocab.special_sep_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_SEP_ID));
     GGUF_GET_KEY(ctx, vocab.special_pad_id, gguf_get_val_u32, GGUF_TYPE_UINT32, false, kv(LLM_KV_TOKENIZER_PAD_ID));
+
+    // build special tokens cache
+    {
+        // TODO: It is unclear (to me) at this point, whether special tokes are guaranteed to be of a deterministic type,
+        //  and will always be correctly labeled in 'added_tokens.json' etc.
+        // The assumption is, since special tokens aren't meant to be exposed to end user, they are designed
+        //  to be unmatchable by the tokenizer, therefore tokens from the vocab, which are unmatchable by the tokenizer
+        //  are special tokens.
+        // From testing, this appears to corelate 1:1 with special tokens.
+        //
+        for (const auto & t: vocab.token_to_id)
```

Comment:
Feel like we have a redundancy here. We have `token_data.type` which is supposed to tell us if a token is special or not. In which cases this wouldn't work?

I guess we can have this piece of code here as a sanity check that the special tokens that we have read from the vocabulary are indeed special, but ideally AFAIU we shouldn't need this code in order to function correctly, correct?

Reply:
Yes. My approach is ment to be temporary ( or a fallback solution ), untill I know for sure special tokens will always be marked as such in the vocab.

@goerch wasn't certain about token types in bpe PR, so I opted for finding special tokens manually for now.

Reply:
Shouldn't we get rid of #3502 first? This looks pretty basic (and terrible :) for me.

Reply:
> Shouldn't we get rid of #3502 first? This looks pretty basic (and terrible :) to me.

I'm open for critique, but you have to clarify what you mean by "this" and "terrible" :) otherwise I'm not sure how to proceed here

Reply:
> We can probably combine both the tokens marked as special and those that are unmatchable by other tokens. And probably print a warning if there is a mismatch between the 2 sets.

Oh I like this idea

Reply:
> > Shouldn't we get rid of #3502 first? This looks pretty basic (and terrible :) to me.
> 
> I'm open for critique, but you have to clarify what you mean by "this" and "terrible" :) otherwise I'm not sure how to proceed here

I meant #3502 looks terrible, sorry (and sorry for being late). @jploski [explained](https://github.com/ggerganov/llama.cpp/issues/2820#issuecomment-1747910558) what we probably need to do for added tokens (i.e. USER_DEFINED as I believe to understand) in the `sentencepiece`-case. I'd start with the idea of replacing `utf8_len` with something like `prefix_match` in `tokenize`:

``` 
        // split string into utf8 chars
        int index = 0;
        size_t offs = 0;
        while (offs < text.size()) {
            llm_symbol sym;
            // size_t len = utf8_len(text[offs]);
            size_t len = prefix_match(text, offs, ...);
            sym.text = text.c_str() + offs;
            sym.n = std::min(len, text.size() - offs);
            offs += sym.n;
            sym.prev = index - 1;
            sym.next = offs == text.size() ? -1 : index + 1;
            index++;
            symbols.emplace_back(sym);
        }
```

and work from there. A similar method could work for GPT2-based tokenizers.

Reply:
Special tokens cache is pretty much done, I added a verification code to compare calculated cache with token types from vocab as per suggestion

I'm currently finishing and testing the special tokens pre-matcher part, I rewrote it for zero string copy, I'll probably push it tomorrow.

After that, we would have special tokens cache and special tokens pre-matcher done, and I can start looking into the actual tokenizer main loop.

So most likely I'll have a solid base ready tomorrow, for further optimisations and improvements to your heart's content :)

Reply:
Just to be sure about our nomenclature, do you mean special tokens == added tokens? I'm mostly leaning towards the nomenclature used in HF's `tokenizer.json` which contains `added_tokens` with the `is_special` attribute. From @jploski I learned that these added tokens are probably very similar to `sentence_piece` `USER_DEFINED` tokens. 

And one more question: do we agree that `vocab.json`, `merges.txt`, `added_tokens.json` and `special_tokens_map.json` are probably older HF (or GPT2) serialization formats and we should find all this information in `tokenizer.json` for newer models (paging @apage43 too because he knows a lot about this stuff)? W.r.t. to these serialization formats we also seem to have some reports indicating that we don't support both of them equally well.

Reply:
> do you mean special tokens == added tokens?

Oh my, you better grab a chair because you are in for an adventure with this question :)

A really condensed version of what I learned so far:

I couldn't find any clear definition or explanation, but from what I found, there seem to be only one way all pieces of this puzzle fall together to form a clear picture.

A vocab, naturally forms a tree like structure ( several trees actually, there is no common root )

Tokenizer does it's thing, by grabbing small portions of the input text, and trying to steal neighboring portions to form a valid token, rinse and repeat, until nothing can be borrowed from either left or right to merge with and form a valid token.

So basically, tokenizer climbs that tokens tree down from single bytes to full tokens, stepping only on valid branches ( tokens )

Applying this idea backwards, by checking the entire vocab token by token, trying to split its text representation in two and checking if it still forms two valid tokens, you can clearly see some of the tokens in the vocab are not part of that tree family, and will never be matched by the tokenizer.

Which happens to match perfectly with **normal** tokens being marked as `LLAMA_TOKEN_TYPE_NORMAL` in the vocab, and tokens not being part of the tokens tree family being marked as whatever type being **not** `LLAMA_TOKEN_TYPE_NORMAL`

Therefore, I'm using the term *special* token, refering to tokens which are "hidden" from the user and un-matchable by tokenizer.

From the practical point of view, that subset of "special" tokens contains exactly what one would expect, `<s>`,`</s>`,`<unk>` and in case of mistral openorca `<|im_start|>`,`<|im_end|>`, AKA tokens which are used to control the flow of inference.

And so from the point of view of actual inference, the tokenizer has to be extended the same way, independently of the actual token type, because only **normal** vs **not normal** token type matter at that point.

Additional distinction between USER_DEFINED, CONTROL etc can and has to be done the same way, with a tokenizer "pre-matcher", at which point per token type decisions or actions are trivial to implement if needed.

Reply:
Thanks for taking the time! My mental image is slightly different: I believe to understand the initial training of the tokenizer leads to an initial vocabulary and later model refinement tries to extend this basic vocabulary with added tokens without retraining the tokenizer on the original dataset. These added tokens should be intercepted before the original tokenizer interpretation. In my mental image special tokens somehow describe the effect these added tokens have on detokenization, i.e. should they be displayed or not for example. But I might be completely off here.

Reply:
I simply do not know if there is any significance to added "normal" tokens ( added in fine tuning etc )

From what I understand, in tokenizer specifically ( ignoring detokenizer ), normal tokens added the way you say, would simply be tokenized one step further into a longer token instead of being represented by a sequence of smaller tokens.

If such added tokens, are not integrated with the tree structure of the vocab, user would not be able to use them
If however those added tokens are properly coherent with the vocab, tokenizer would match them already, as it is

Which to me sort of seems to defy the purpose of adding "normal" tokens this way

I might be wrong about this part, but that still means a "tokenizer preprocessor" of some sort is needed, and this PR fits that role

Reply:
> I simply do not know if there is any significance to added "normal" tokens ( added in fine tuning etc )

I believe they have to be processed before the core tokenizer routine in something like `prefix_match` (only exception are the already defined core tokenizer CONTROL tokens). `mpt` added tokens for example (from `tokenizer.json`):

```    
    {
      "id": 0,
      "content": "<|endoftext|>", <-- core, i.e CONTROL
      "single_word": false,
      "lstrip": false,
      "rstrip": false,
      "normalized": false,
      "special": true
    },
    {
      "id": 1,
      "content": "<|padding|>", <-- unsure about this because it is a valid GPT2 token AFAIU
      "single_word": false,
      "lstrip": false,
      "rstrip": false,
      "normalized": false,
      "special": true
    },
    {
      "id": 50254,
      "content": "                        ",  <-- extension, i.e USER_DEFINED
      "single_word": false,
      "lstrip": false,
      "rstrip": false,
      "normalized": true,
      "special": false
    },
    [...]
```